data/src/main/java/com/microsoft/azure/kusto/data/auth/ConnectionStringBuilder.java

```diff
@@ -54,6 +54,27 @@ private ConnectionStringBuilder(String clusterUrl) {
         this.applicationNameForTracing = null;
     }
 
+    public ConnectionStringBuilder(ConnectionStringBuilder other){
+        this.clusterUrl = other.clusterUrl;
+        this.usernameHint = other.usernameHint;
+        this.applicationClientId = other.applicationClientId;
+        this.applicationKey = other.applicationKey;
+        this.x509Certificate = other.x509Certificate;
```

Comment:
I'm not sure it's necessary here because CSBs have such a short useful lifetime (just for creating the client), but nevertheless I'd deep copy mutable objects here to prevent hard-to-find and unintended bugs where changing 1 CSB could affect the other.

Reply:
The potential mutable objects are:
```
private X509Certificate x509Certificate;
private List<X509Certificate> x509CertificateChain;
private PrivateKey privateKey;
private Callable<String> tokenProvider;
```

Other than list, I don't know how to clone the rest

Reply:
I think `this.x509Certificate = SerializationUtils.clone(other.x509Certificate);` would work.

Reply:
Not sure I'm comfortable with it
What if there are fields that can't be serialized?
I think it would lead to subtle errors

Reply:
The method signature is:
`public static <T extends Serializable> T clone(final T object)`

It indeeds requires that passed parameters are serializable, but the good news is that it throws `SerializationException` if it can't, so any errors should be easy to identify.

Of course any subtype can be passed, but `X509Certificate` itself seems `Serializable`, because its parent `Certificate` is, and it doesn't contain any additional members that can be set. (It has 2 transient members, but those are simply from the parent object)
And `PrivateKey` implements `Key`, which is serializable.

We can't really deep-copy `Callable`.

Reply:
Actually, I see that `X509Certificate` and `PrivateKey` seem to effectively be immutable, as neither has methods to change their state (except `destroy()`, but arguably that should be allowed to destroy).

So nevermind, there isn't a reasonable need to deep-copy these 2.

Reply:
So can I close it?